### PR TITLE
[Backport] Improve build configuration time

### PR DESCRIPTION
**Pull request by mark-vieira**
Backport of #41392

### CODE REVIEW

**Comment by mark-vieira** [comment]
@atorok @rjernst So I've backported these changes to `7.x`. This really accomplishes 2 things.

1. We get the config time bonus in `7.x`.
2. It makes backporting other build changes, especially those that touch `BuildPlugin` a lot easier since this is a very large change.

The bummer with 7.x is that it still supports the JDK 8 runtime so there are a shit load of conditionals around runtime java version, which now is deferred until execution time. We can either solve this with the eventual goal to completely defer JDK provisioning to the JDK plugin, or just say screw it, and leave it as-is.

**Comment by atorok** [comment]
Thanks! We really need to be able to backport build changes so this was a must. 
I think that provisioning with the JDK plugin is a good idea which we should do regardless, not sure if this PR should influence the priority of doing so unless we discover all those conditionals to get in the way. I'm assuming they won't do so much with most development happening on master.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy [hunk 1]
```diff
@@ -927,19 +739,24 @@ class BuildPlugin implements Plugin<Project> {
                     jarTask.manifest.getAttributes().clear()
                 }
             }
+
             // add license/notice files
             project.afterEvaluate {
-                if (project.licenseFile == null || project.noticeFile == null) {
+                if (ext.has('licenseFile') == false || ext.get('licenseFile') == null || ext.has('noticeFile') == false || ext.get('noticeFile') == null) {
```
Comment (synhershko) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: This seems to have broken our custom plugin builds. The same build.gradle file which worked for all 6.x and 7.x builds with esplugin is now broken and fails with

```
* What went wrong:
A problem occurred configuring root project 'elasticsearch-analysis-hebrew'.
> Failed to notify project evaluation listener.
   > Must specify license and notice file for project :
   > Must specify license and notice file for project :
   > Must specify license and notice file for project :
```

It is indeed possible that I'm missing something, but we do have noticeFile and licenseFile properly declared and moving them from project to ext to esplugin doesn't work.

Any idea?

Comment (synhershko) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: OK n/m just found https://github.com/elastic/elasticsearch/issues/45073#issuecomment-517388642 , which fixes this issue

**Pull request merged**